scopes/preview/ui/component-preview/preview.tsx

```diff
@@ -43,17 +49,64 @@ export function ComponentPreview({
   component,
   previewName,
   queryParams,
-  pubsub = true,
+  pubsub,
   fullContentHeight = false,
   style,
   ...rest
 }: ComponentPreviewProps) {
-  const [iframeRef, iframeHeight] = useIframeContentHeight({ skip: !fullContentHeight });
+  // const [iframeRef, iframeHeight] = useIframeContentHeight({ skip: !fullContentHeight });
+  const iframeRef = useRef<HTMLIFrameElement>(null);
+  const [width, setWidth] = useState(0);
+  const [height, setHeight] = useState(0);
+  const containerRef = useRef<HTMLDivElement>(null);
   // @ts-ignore (https://github.com/frenic/csstype/issues/156)
-  const height = iframeHeight || style?.height;
+  // const height = iframeHeight || style?.height;
   usePubSubIframe(pubsub ? iframeRef : undefined);
+  // const pubsubContext = usePubSub();
+  // pubsubContext?.connect(iframeHeight);
+  useEffect(() => {
+    if (!iframeRef.current || pubsub) return;
+    connectToChild({
+      iframe: iframeRef.current,
+      methods: {
+        pub: (event, message) => {
+          if (message.type === 'preview-size') {
+            setWidth(message.data.width);
+            setHeight(message.data.height);
+          }
+        },
+      },
+    });
+  });
+  const isScaling = component.preview?.isScaling;
+
+  const params = Array.isArray(queryParams)
+    ? queryParams.concat('viewport=1280')
+    : compact([queryParams, 'viewport=1280']);
 
-  const url = toPreviewUrl(component, previewName, queryParams);
+  const url = toPreviewUrl(component, previewName, isScaling ? params : queryParams);
+  // const currentHeight = fullContentHeight ? '100%' : height || 1024;
+  const containerWidth = containerRef.current?.offsetWidth || 0;
+  const currentWidth = fullContentHeight ? '100%' : width || 1280;
+  const targetWidth = currentWidth < containerWidth ? containerWidth : currentWidth;
+  const targetHeight = height !== 0 ? height : 5000;
 
-  return <iframe {...rest} ref={iframeRef} style={{ ...style, height }} src={url} />;
+  return (
+    <div ref={containerRef}>
+      <iframe
+        {...rest}
+        ref={iframeRef}
+        style={{
+          ...style,
+          height: isScaling ? targetHeight : '100%',
+          width: isScaling ? targetWidth : '100%',
+          visibility: width === 0 && isScaling && !fullContentHeight ? 'hidden' : undefined,
+          transform: fullContentHeight ? '' : computePreviewScale(width, containerWidth),
```

Comment:
its kinda difficult to track what is supposed to happen and when. we might want to move the calculations into a separate function